tests/results/test_result_fetch.py

```diff
@@ -0,0 +1,137 @@
+import pytest
+
+from prefect import flow, task
+from prefect.client.schemas import Completed
+from prefect.results import ResultLiteral
+from prefect.settings import PREFECT_ASYNC_FETCH_STATE_RESULT, temporary_settings
+
+
+@pytest.fixture(autouse=True)
+def disable_fetch_by_default():
+    """
+    The test suite defaults to the future behavior.
+
+    For these tests, we enable the default user behavior.
+    """
+    with temporary_settings({PREFECT_ASYNC_FETCH_STATE_RESULT: False}):
+        yield
+
+
+async def test_async_result_raises_deprecation_warning():
+    # This test creates a state directly because a flows do not yet return the new
+    # result types
+    state = Completed(data=await ResultLiteral.create(True))
+    result = state.result(fetch=False)
+
+    with pytest.warns(
+        DeprecationWarning,
+        match=r"State.result\(\) was called from an async context but not awaited.",
+    ):
+        result = state.result()
+
+    # A result type is returned
+    assert isinstance(result, ResultLiteral)
+    assert await result.get() is True
+
+
+async def test_async_result_warnings_are_not_raised_by_engine():
+    # Since most of our tests are run with the opt-in globally enabled, this test
+    # covers a bunch of features to cover remaining cases where we may internally
+    # call `State.result` incorrectly.
```

Comment:
Not really enthused about this kind of test but it'll be useful as I update the engine to use the new result types.